### ISSUE TITLE
 stream_socket_client() error for Apple Push

### ISSUE THREAD

**Issue by lorenzosfarra** [Ph3nol/NotificationPusher#102]
Hello everyone,

the problem is strange. I have seen some questions about it but according to my tests I am not able to solve.

The error is the following:

```
Warning: stream_socket_client(): Failed to enable crypto in /home/wineowine/stage/NotificationPusher/vendor/zendframework/zendservice-apple-apns/library/ZendService/Apple/Apns/Client/AbstractClient.php on line 102

Warning: stream_socket_client(): unable to connect to tls://gateway.push.apple.com:2195 (Unknown error) in /home/wineowine/stage/NotificationPusher/vendor/zendframework/zendservice-apple-apns/library/ZendService/Apple/Apns/Client/AbstractClient.php on line 102

Fatal error: Uncaught exception 'ZendService\Apple\Exception\RuntimeException' with message 'Unable to connect: tls://gateway.push.apple.com:2195: 0 ()' in /home/wineowine/stage/NotificationPusher/vendor/zendframework/zendservice-apple-apns/library/ZendService/Apple/Apns/Client/AbstractClient.php:105 Stack trace: #0 /home/wineowine/stage/NotificationPusher/vendor/zendframework/zendservice-apple-apns/library/ZendService/Apple/Apns/Client/AbstractClient.php(78): ZendService\Apple\Apns\Client\AbstractClient->connect('tls://gateway.p...', Array) #1 /home/wineowine/stage/NotificationPusher/src/Sly/NotificationPusher/Adapter/Apns.php(120): ZendService\Apple\Apns\Client\AbstractClient->open(1, '/etc/ssl/localc...', NULL) #2 /home/wineowine/stage/NotificationPusher/src/Sly/NotificationPusher/Adapter/Apns.php(133): Sly\NotificationPusher\Adapter\Apns->getOpenedClient(Object(ZendService\Apple\Apns\Client\Message)) #3 /home/wineowine/stage/NotificationPusher/src/Sly/NotificationPusher/Adapter/Apns.php(68): Sly\NotificationPusher\Ada in /home/wineowine/stage/NotificationPusher/vendor/zendframework/zendservice-apple-apns/library/ZendService/Apple/Apns/Client/AbstractClient.php on line 105
```

I have, of course, tried to exclude that there is something related to the connection from my server to the Apple server, but it's fine:

```
telnet gateway.push.apple.com 2195
Trying 17.143.162.139...
Connected to gateway.push-apple.com.akadns.net.
Escape character is '^]'.
```

And I have tried to check my certificate. I will hide all the output but the relevant part is that it says:

```
    Start Time: 1455784403
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
```

What else I could check? 

Basically the code is:

```
    $pushManager = new PushManager(PushManager::ENVIRONMENT_PROD);
    // Then declare an adapter.
    $apnsAdapter = new ApnsAdapter(array(
        'certificate' => $certificate
    ));
    [...skipping stuff...]
    // Load the devices
    $devices = new Sly\NotificationPusher\Collection\DeviceCollection($iOSDevicesTokenList);
    // Load the message
    $message = new Sly\NotificationPusher\Model\Message('title',
      array('custom' => $dictionary)
    );
    // Finally, create and add the push to the manager, and push it!
    $push = new Push($apnsAdapter, $devices, $message);
    $pushManager->add($push);
    $pushResult = $pushManager->push();
```

Any hints?

Thanks,
  Lorenzo

**Comment by lorenzosfarra** [comment]
It's strange, at the end I gave up and I've tried to generate new certificates.
Now in the Apple web interface I have a general "sandbox and production" certificate, I have tried with that, exported as a .pem and created a symlink (both for the previous sandbox and production certificates) that points to the new one, and everything works fine.
Maybe I have missed some communications by Apple?
The previous certificates were working fine, and they are still valid, so I don't know why they were not working anymore.

Thanks,
  Lorenzo

**Comment by jamrader** [comment]
I'm getting the same "unable to connect to tls://gateway.push.apple.com:2195," but I assume my pem file is fine, because I just used the p12 key to send a push via another push service.

**Comment by lorenzosfarra** [comment]
You can check your .pem file with 
```
openssl s_client -connect gateway.sandbox.push.apple.com:2195 -cert <your_file>.pem -debug -showcerts -CAfile /path/to/entrust_ev_ca.cer
```
(this is for sandbox, change the URL for the production environment).

In my case (Ubuntu 14.04) the last certificate is located here: `/etc/ssl/certs/Entrust_Root_Certification_Authority.pem`

**Comment by jamrader** [comment]
Thanks. The problem was port 2195 was blocked.

**Comment by sandy-abrightmind** [comment]
How does the pem file get the Passphrase?

**Comment by seyfer** [comment]
@lorenzosfarra, seems like it was the issue connected with an expired certificate or closed port.
